### PR TITLE
fix: use addOpenedChangeListener instead of subscribing to opened-changed

### LINKED ISSUE #4743: Dialog is not detached if another modal is opened during "close"
### Description

I have a `Dialog` which is auto added to the `UI` on open.
A registered `OpenedChangedListener` opens a new modal dialog when the other dialog gets closed.
Now I noticed, that the first dialog is closed, but not detached from the `UI`.

### Expected outcome

As the dialog has been added to the `UI` automatically, it should also be detached automatically.

### Minimal reproducible example

```
{
	TextField name = new TextField("Your name");
	Button sayHello = new Button("Say hello");
	sayHello.addClickListener(e -> {
		Dialog dlg = new Dialog();
		dlg.add(new VerticalLayout(
			new Span("Hello " + name.getValue()),
			new Button("Close", ev -> dlg.close())
		));
		dlg.addOpenedChangeListener(ev -> {
			if (!ev.isOpened()) {
				Dialog dlg2 = new Dialog();
				dlg2.add(new Span("Inner"));
				dlg2.open();
			}
		});
		dlg.addDetachListener(ev -> System.out.println("Detached"));
		dlg.getElement().addEventListener("opened-changed", event -> System.out.println(dlg.isOpened() ? "Open" : "Close"));
		dlg.open();
	});
	setMargin(true);
	setVerticalComponentAlignment(Alignment.END, name, sayHello);
	add(name, sayHello);
}
```

### Steps to reproduce

1. Download a new Vaadin starter project
2. Copy the above code into the constructor of the `MainView`.
3. Run the project
4. Click the "Say hello" button
5. Click the "Close" button in the dialog

The console should print:
- Open
- Detach
- Close

But it only prints "Open", detach and the `opened-hanged`-listener have not been called.

### Environment

Vaadin: 23.3.6
Flow: 23.3.4
Java: Eclipse Adoptium 17.0.6
OS: amd64 Windows 11 10.0

### Browsers

Issue is not browser related

### Notes
A dialog, that is auto added to the `UI` gets detached after a roundtrip to the client:
1. Server sends opened = false to the client
2. Client sends a opened-change event to the server
3. Server has a opened-changed listener, that detaches an auto added dialog

Unfortunately the new modal dialog makes the original dialog node inert and therefor the opened-changed listener is not called.

### PR DESCRIPTION
**Pull request by ugur-vaadin**
## Description

When a new `Dialog` is opened before a closing `Dialog` gets detached, it is set inert via the parent. Therefore, it does not get closed and detached properly. 

This PR makes the `Dialog` use `addOpenedChangeListener` instead of subscribing to `opened-changed` event for removing the `Dialog` from its parent. This way, server side `Dialog.close()` calls can be handled properly instead of waiting for the client side `opened-change` event.

This PR should not be merged until the Dialog test refactoring [PR](https://github.com/vaadin/flow-components/pull/4850) is merged.

Fixes #4743 and #4026  

## Type of change

- [x] Bugfix
- [ ] Feature

## Checklist

- [x] I have read the contribution guide: https://vaadin.com/docs/latest/contributing/overview
- [x] I have added a description following the guideline.
- [x] The issue is created in the corresponding repository and I have referenced it.
- [x] I have added tests to ensure my change is effective and works as intended.
- [x] New and existing tests are passing locally with my change.
- [x] I have performed self-review and corrected misspellings.

### CODE REVIEW

File: vaadin-dialog-flow-parent/vaadin-dialog-flow/src/main/java/com/vaadin/flow/component/dialog/Dialog.java [hunk 1]
```diff
@@ -801,6 +803,11 @@ public void setOpened(boolean opened) {
             ensureAttached();
         }
         setModality(opened && isModal());
+        // This part makes sure that the dialog gets detached if it gets closed
+        // but is set inert before it can get detached.
+        if (!opened) {
+            ElementUtil.setIgnoreParentInert(getElement(), true);
```
Comment (vursen) [inline vaadin-dialog-flow-parent/vaadin-dialog-flow/src/main/java/com/vaadin/flow/component/dialog/Dialog.java:809]: I have a concern that the current approach allows unintentionally making the dialog ignore the parent inert when not supposed. For example, when you call `setOpened(false)` on a closed dialog.

Comment (ugur-vaadin) [reply vaadin-dialog-flow-parent/vaadin-dialog-flow/src/main/java/com/vaadin/flow/component/dialog/Dialog.java:809]: I added `isOpened()` to the if check to ensure that this line is only executed when an open dialog is being closed.

File: vaadin-dialog-flow-parent/vaadin-dialog-flow-integration-tests/src/test/java/com/vaadin/flow/component/dialog/tests/SubDialogOpenedOnOpenedChangeIT.java [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+    @Test
+    public void openMainDialog_openSubDialog_mainDialogGetsDetached() {
+        waitForElementPresent(By.id("open-main-dialog"));
+        findElement(By.id("open-main-dialog")).click();
+
+        waitForElementPresent(By.id("close-main-dialog-and-open-sub-dialog"));
+        findElement(By.id("close-main-dialog-and-open-sub-dialog")).click();
+
+        WebElement output = findElement(By.id("output"));
+        Assert.assertEquals("Detached", output.getText());
```
Comment (vursen) [inline vaadin-dialog-flow-parent/vaadin-dialog-flow-integration-tests/src/test/java/com/vaadin/flow/component/dialog/tests/SubDialogOpenedOnOpenedChangeIT.java:43]: **nit:**

```suggestion
        $("button").id("open-main-dialog").click();
        $("button").id("close-main-dialog-and-open-sub-dialog").click();

        WebElement output = findElement(By.id("output"));
        Assert.assertEquals("Detached", output.getText());
```

File: vaadin-dialog-flow-parent/vaadin-dialog-flow-integration-tests/src/main/java/com/vaadin/flow/component/dialog/tests/SubDialogOpenedOnOpenedChangeView.java [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+
+/**
+ * Page created for testing purposes. Not suitable for demos.
+ *
+ * @author Vaadin Ltd
+ */
+@Route("vaadin-dialog/sub-dialog-opened-on-opened-change")
+public class SubDialogOpenedOnOpenedChangeView extends Div {
+
+    public SubDialogOpenedOnOpenedChangeView() {
```
Comment (vursen) [inline vaadin-dialog-flow-parent/vaadin-dialog-flow-integration-tests/src/main/java/com/vaadin/flow/component/dialog/tests/SubDialogOpenedOnOpenedChangeView.java:32]: **nit:**

```suggestion
    public OpeningSubDialogOnOpenedChangeView() {
```

File: vaadin-dialog-flow-parent/vaadin-dialog-flow-integration-tests/src/main/java/com/vaadin/flow/component/dialog/tests/SubDialogOpenedOnOpenedChangeView.java [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+package com.vaadin.flow.component.dialog.tests;
+
+import com.vaadin.flow.component.button.Button;
+import com.vaadin.flow.component.dialog.Dialog;
+import com.vaadin.flow.component.html.Div;
+import com.vaadin.flow.component.html.Span;
+import com.vaadin.flow.router.Route;
+
+/**
+ * Page created for testing purposes. Not suitable for demos.
```
Comment (vursen) [inline vaadin-dialog-flow-parent/vaadin-dialog-flow-integration-tests/src/main/java/com/vaadin/flow/component/dialog/tests/SubDialogOpenedOnOpenedChangeView.java:25]: I don't think this comment is relevant.

```suggestion
```

**Review by vursen — APPROVED** [review]

**Pull request merged**